### PR TITLE
Treat and test all bonded interactions in the TPR parser

### LINKED ISSUE #222: TPRParser not detecting bonds.
When loading with a .tpr as the topology bond information isn't taken from the tpr.

I've attached a simple system, just a box of flexible water.

The angles seem to be loaded fine, just not the bonds?  `gmxdump` tells me that the tpr has the bonds.

I've tried rolling back to the 0.8.1 release and bonds still aren't in _psf so I don't think we broke it recently with the class based topologies.

There's also not any testing for the topology loading in tpr files.

Original issue reported on code.google.com by `richardjgowers` on 9 Mar 2015 at 11:17

Attachments:
* [topol.tpr](https://storage.googleapis.com/google-code-attachments/mdanalysis/issue-222/comment-0/topol.tpr)
* [out.gro](https://storage.googleapis.com/google-code-attachments/mdanalysis/issue-222/comment-0/out.gro)

### PR DESCRIPTION
**Pull request by jbarnoud**
Working on #458, I realize that very few bonded interactions available in Gromacs were actually used by the TPR parser. Most notably, the constraints are not treated as bonds leading to issues #222 and #352.

Here, I make the TPR parser treat all the bonded interactions described in table 5.5 of the Gromacs PDF manual. I also added tests to make sure each interaction is treated as expected.

The bonded interactions taken into account are:

* bonds: regular bonds (type 1), G96 bonds (type 2), Morse (type 3), cubic bonds (type 4), connections (type 5), harmonic potentials (type 6), FENE bonds (type 7), restraint potentials (type 10), tabulated potential with exclusion/connection (type 8), tabulated potential without exclusion/connection (type 9), constraints with exclusion/connection (type 1), constraints without exclusion/connection (type 2)
* angles: regular angles (type 1), G96 angles (type 2), cross bond-bond (type3), cross-bond-angle (type 4), Urey-Bradley (type 5), quartic angles (type 6), restricted bending potential (type 10), tabulated angles (type 8)
* dihedrals: proper dihedrals (type 1 and type 9), Ryckaert-Bellemans dihedrals (type 3), Fourier dihedrals (type 5), restricted dihedrals (type 10), combined bending-torsion potentials (type 11), tabulated dihedral (type 8)
* impropers: improper dihedrals (type 2), periodic improper dihedrals (type 4)

I am not familiar with all these interactions so I may have classed some improperly. I may have omited some too.

Prior to this commit, a comment read:

> this is possible a bug in gromacs, the so-named Proper Dih is actually Improper Dih

I did not find other mention of thid issue, nor did I noticed it from a basic test. Therefore, with this commit, proper dihedrals are treated as such. This may be a mistake, though.

Also prior to this commit, 3 body virtual sites were treated as improper dihedral. They are not anymore.

The tests make sure that the TPR parser reads all the possible bonded interactions. The files in `testsuite/MDAnalysisTests/data/tprs/all_bonded` describe a system that is supposed to contain all these bonded interactions.

When a new bonded interaction is added to Gromacs, it should also be added to `testsuite/MDAnalysisTests/data/tprs/all_bonded/dummy.top`, and to `testsuite/MDAnalysisTests/datafiles.py`. Do not forget to rebuild `testsuite/MDAnalysisTests/data/tprs/all_bonded/dummy_5.1.tpr`.

This PR should fix #222 and #352.

### CODE REVIEW

File: testsuite/setup.py [hunk 1]
```diff
@@ -102,7 +102,7 @@
           package_data={'MDAnalysisTests':
               [
                   'data/*.psf', 'data/*.dcd', 'data/*.pdb',
-                  'data/tprs/*.tpr', 'data/*.tpr',
+                  'data/tprs/*.tpr', 'data/tprs/all_bonded/*.tpr', 'data/*.tpr',
```
Comment (richardjgowers) [inline testsuite/setup.py]: I think you'll need to add the gro, top and mdp files here too

**Comment by richardjgowers** [comment]
So this isn't really the issue, but how tricky would it be to pull out the coordinate information from the TPR?

File: testsuite/MDAnalysisTests/test_tprparser.py [hunk 2]
```diff
@@ -236,3 +238,64 @@ class TPR510(TPRBase):
 
 class TestTPR510(_TestTopology, TPR510):
     """Testing TPR version 103"""
+
+
+def _test_is_in_topology(name, elements, topology_section, topology_path):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/test_tprparser.py]: If you've gone to the effort of writing down all the bond names and passing them this far, you may as well include some sort of detailed error message, `"Bond type {} not found".format(name)`

**Comment by d-beltran** [comment]
Hi everyone,

I'm having problems with MDAnalysis considering as bonds some restrains I am not interested in and it is not easy to tell the difference between covalent bonds and restrains (automatically). The 'bonds.types' section returns a list filled with None and, to my knowledge, there is no way to tell the parser to ignore specific bond types.

Is there a way to get only the covalent bonds?

Best regards and thank you for maintaining MDAnalysis :)

**Pull request merged**